### PR TITLE
i#4134: Add sample to highlight drbbdup API usage

### PR DESCRIPTION
**Pull request by johnfxgalea**
Includes a sample to demonstrate how to use drbbdup.

The client sets up two instrumentation cases per basic block. One case tracks the hit count of  a basic block when _cold_ to determine whether it exceeds a threshold and should be considered _hot_. The second version simply counts the number of times a hot basic block is executed.

Issue: #4134

### CODE REVIEW

File: api/samples/hot_bbcount.c [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+
+static void
+event_exit(void)
+{
+#ifdef SHOW_RESULTS
+    char msg[512];
+    int len;
+    len = dr_snprintf(msg, sizeof(msg) / sizeof(msg[0]),
+                      "Instrumentation results:\n"
+                      "%10d basic block executions\n",
```
Comment (derekbruening) [inline api/samples/hot_bbcount.c]: This should say "hot" basic block

File: api/samples/CMakeLists.txt [hunk 2]
```diff
@@ -236,6 +236,7 @@ if (X86) # FIXME i#1551, i#1569: port to ARM and AArch64
   add_sample_client(instrcalls  "instrcalls.c;utils.c"  "drmgr;drsyms;drx")
   # dr_insert_cbr_instrument_ex is NYI
   add_sample_client(cbrtrace    "cbrtrace.c;utils.c"    "drmgr;drx")
+  add_sample_client(hot_bbcount "hot_bbcount.c"         "drmgr;drreg;drbbdup;drx")
```
Comment (derekbruening) [inline api/samples/CMakeLists.txt]: Please add an entry to samples.dox too

File: api/samples/hot_bbcount.c [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+/* Start counting once a bb has been executed at least 10 times. */
+#define HIT_THRESHOLD 1000
+
+#ifdef WINDOWS
+#    define DISPLAY_STRING(msg) dr_messagebox(msg)
+#else
+#    define DISPLAY_STRING(msg) dr_printf("%s\n", msg);
+#endif
+
+#define NULL_TERMINATE(buf) buf[(sizeof(buf) / sizeof(buf[0])) - 1] = '\0'
```
[the remark this answers is not in the document]
Comment (johnfxgalea) [reply api/samples/hot_bbcount.c]: I think we should add a separate issue for this change. The required mod is not unique to this sample. WDYT?

File: api/samples/hot_bbcount.c [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+    dr_register_exit_event(event_exit);
+
+    hashtable_init_ex(&hit_count_table, HASH_BITS, HASH_INTPTR, false /*!strdup*/,
+                      false /*synchronization is external*/, destroy_hit_count, NULL,
+                      NULL);
+
+    /* Initialise drbbdup. Essentially, drbbdup requires
+     * the client to pass a set of call-back functions.
+     */
+    drbbdup_options_t opts = { sizeof(drbbdup_options_t),
```
[the remark this answers is not in the document]
Comment (johnfxgalea) [reply api/samples/hot_bbcount.c]: Done.

File: api/samples/hot_bbcount.c [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+
+    /* Set the default case encoding for tracking the hit count of basic blocks. */
+    return 0; /* cold */
+}
+
+static void
+analyse_orig_bb(void *drcontext, void *tag, instrlist_t *bb, void *user_data,
+                IN void **orig_analysis_data)
+{
+    /* Extract bb_pc and set store it as analysis data. */
```
[the remark this answers is not in the document]
Comment (johnfxgalea) [reply api/samples/hot_bbcount.c]: Done.

File: api/samples/hot_bbcount.c [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+
+static uintptr_t
+set_up_bb_dups(void *drbbdup_ctx, void *drcontext, void *tag, instrlist_t *bb,
+               bool *enable_dups, void *user_data)
+{
+    /* Init hit count. */
+    app_pc bb_pc = instr_get_app_pc(instrlist_first_app(bb));
+    hashtable_lock(&hit_count_table);
+    if (hashtable_lookup(&hit_count_table, bb_pc) == NULL) {
+        /* If hit count is not mapped to this bb, then add a new count to the table */
```
[the remark this answers is not in the document]
Comment (johnfxgalea) [reply api/samples/hot_bbcount.c]: Done.

File: api/samples/hot_bbcount.c (hunk 1 above)
Comment (johnfxgalea) [reply api/samples/hot_bbcount.c]: Done.

File: api/samples/CMakeLists.txt (hunk 2 above)
Comment (johnfxgalea) [reply api/samples/CMakeLists.txt]: Oops, forgot that! Done.

File: api/samples/hot_bbcount.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (johnfxgalea) [reply api/samples/hot_bbcount.c]: I added  #4160 so that this can be applied to all samples (including for the drbbdup sample).

**Pull request merged**